### PR TITLE
sbt-ci-release Fixes #196

### LINKED ISSUE #196: Snapshot artifacts published as releases
Artefacts created from `master` are not releases. Although the tests passed, they might contain code not ready for production use. And if we want to publish *actual* release artefacts, there should be a distinction.

Unfortunately artefacts created from `master` are published to both
 * https://oss.sonatype.org/content/repositories/releases/org/scalaz/scalaz-zio_2.12/
 * https://repo1.maven.org/maven2/org/scalaz/scalaz-zio_2.12/

repositories are dedicated only to release artefacts.
The [maven metadata](https://oss.sonatype.org/content/repositories/releases/org/scalaz/scalaz-zio_2.12/maven-metadata.xml) deceivingly suggest that `0.1.0-1167e15` is not only the latest, but also a release.
```xml
<metadata>
  ...
  <versioning>
    <latest>0.1.0-1167e15</latest>
    <release>0.1.0-1167e15</release>
    <versions>
      ...
      <version>0.1.0-1167e15</version>
    </versions>
    ...
  </versioning>
</metadata>
```

This breaks [Ivy versioning semantics in dependency resolution](https://ant.apache.org/ivy//history/latest-milestone/ivyfile/dependency.html), i. e. `latest.release`.

What prohibits from publishing snapshot artefacts to https://oss.sonatype.org/content/repositories/snapshots/org/scalaz/scalaz-zio_2.12/ ?

### PR DESCRIPTION
**Pull request by sideeffffect**
please
 * add a tag to `master` (`v0.1.0`?)
 * add variables `PGP_PASSPHRASE` and `PGP_SECRET` to Travis as described [here](https://github.com/olafurpg/sbt-ci-release#travis)

before merging this to `master`

### CODE REVIEW

**Comment by NeQuissimus** [comment]
We already have `PGP_PASSPHRASE` and I can add the secret tomorrow. It is too large to place in an ENV variable, so I will make an encrypted file. You will need to add this:

```diff
diff --git a/.travis.yml b/.travis.yml
index f43c737..6fc4789 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -11,6 +11,8 @@ install:
 jdk:
 - oraclejdk8
 before_install:
+- openssl aes-256-cbc -K $encrypted_316f730cabee_key -iv $encrypted_316f730cabee_iv
+  -in ./project/secret -out ./project/zecret -d
 - export PATH=${PATH}:./vendor/bundle
 script:
 - sbt ++$TRAVIS_SCALA_VERSION scalafmtCheck test:scalafmtCheck scalafmtSbtCheck coverage
```

**Comment by sideeffffect** [comment]
Travis is misconfigured :(
does somebody understand why Travis doesn't run `TEST="scalafmt"` and `TEST="compile, test and microsite"` for `2.11.12`? Instead it runs just dumb `sbt test` -- where does it take the idea to do that???

**Comment by ktonga** [comment]
Maybe b/c it's not picking up your dank config and that's the default for `language: scala`

File: .travis.yml [hunk 1]
```diff
@@ -1,26 +1,49 @@
...
+
 install:
 - rvm use 2.4.4 --install --fuzzy
 - gem update --system
 - gem install sass
 - gem install jekyll -v 3.8.2
-jdk:
-- oraclejdk8
+
+stages:
```
[the remark this answers is not in the document]
Comment (sideeffffect) [reply .travis.yml]: not in the slightest :D I'll just have to bite my way through it with trial, error, googling and docs reading

[the remark this answers is not in the document]
Comment (NeQuissimus) [reply .travis.yml]: It is probably best to look at what Scalaz 7.3 does (or the way zio `master` currently is). At least, that works :)

File: .travis.yml [hunk 2]
```diff
@@ -1,26 +1,48 @@
...
+#  if: (branch = master AND type = push) OR (tag IS present)
+
+matrix:
+  include:
+  - env: TEST="scalafmt"
+    script: sbt ++$TRAVIS_SCALA_VERSION scalafmtCheck test:scalafmtCheck scalafmtSbtCheck
+  - env: TEST="test"
+    script: sbt ++$TRAVIS_SCALA_VERSION coverage test coverageReport microsite/makeMicrosite && bash <(curl -s https://codecov.io/bash)
+  - stage: release
+    script: sbt ci-release
```
Comment (NeQuissimus) [inline .travis.yml]: We need releases for both 2.11 and 2.12, `++$TRAVIS_SCALA_VERSION` should always be included

File: .travis.yml [hunk 3]
```diff
@@ -1,26 +1,48 @@
...
 - gem update --system
 - gem install sass
 - gem install jekyll -v 3.8.2
-jdk:
-- oraclejdk8
+
+stages:
+- name: test
+- name: release
+#  if: (branch = master AND type = push) OR (tag IS present)
```
Comment (NeQuissimus) [inline .travis.yml]: We are going to need something like this :)

File: build.sbt [hunk 4]
```diff
@@ -3,45 +3,23 @@ import sbtcrossproject.CrossPlugin.autoImport.crossProject
...
-  .getOrElse(baseDirectory.value / "project" / "secring.gpg")
-pgpPublicRing in Global := sys.env
-  .get("PGP_PUBRING")
-  .map(file)
-  .getOrElse(baseDirectory.value / "project" / "pubring.gpg")
-useGpg in Global := false
-pgpSigningKey in Global := Some(-6073725311241139476L) // "ABB5CB02682C4AEC"
-
-credentials in ThisBuild ++= sonataCredentials.toSeq
+pgpSecretRing in Global := baseDirectory.value / "project" / "zecret"
```
Comment (NeQuissimus) [inline build.sbt]: I don't think this is the secret ring. I think you need to load this file into the environment variable `PGP_SECRET`.

File: .travis.yml [hunk 5]
```diff
@@ -1,26 +1,59 @@
...
+    env: TEST="test"
+    script: sbt ++$TRAVIS_SCALA_VERSION coverage test coverageReport microsite/makeMicrosite && bash <(curl -s https://codecov.io/bash)
+  - jdk: oraclejdk8
+    scala: 2.12.6
+    env: TEST="test"
+    script: sbt ++$TRAVIS_SCALA_VERSION coverage test coverageReport microsite/makeMicrosite && bash <(curl -s https://codecov.io/bash)
+  - stage: release
+    jdk: oraclejdk8
+    scala: 2.11.12
+    script: sbt ci-release
```
Comment (NeQuissimus) [inline .travis.yml]: You need `++$TRAVIS_SCALA_VERSION` to be part of the ci-release

File: build.sbt (hunk 4 above)
Comment (sideeffffect) [reply build.sbt]: any idea how to do that?

File: .travis.yml (hunk 5 above)
Comment (sideeffffect) [reply .travis.yml]: oh, thanks!

File: .travis.yml (hunk 3 above)
Comment (sideeffffect) [reply .travis.yml]: I was just playing around :)

File: build.sbt (hunk 4 above)
Comment (NeQuissimus) [reply build.sbt]: Execute a script?
`export PGP_SECRET=$(<./project/zecret)`

File: .travis.yml [hunk 6]
```diff
@@ -1,26 +1,59 @@
...
+    scala: 2.12.6
+    env: TEST="test"
+    script: sbt ++$TRAVIS_SCALA_VERSION coverage test coverageReport microsite/makeMicrosite && bash <(curl -s https://codecov.io/bash)
+  - stage: release
+    jdk: oraclejdk8
+    scala: 2.11.12
+    script: sbt ++$TRAVIS_SCALA_VERSION ci-release
+  - stage: release
+    jdk: oraclejdk8
+    scala: 2.12.6
```
Comment (NeQuissimus) [inline .travis.yml]: You'll want to put the scala versions back up top

Comment (sideeffffect) [reply .travis.yml]: why? having them there was the root of the problem I had with misconfigured Travis

Comment (NeQuissimus) [reply .travis.yml]: hmmm, seems strange that would be the issue...

**Comment by sideeffffect** [comment]
@NeQuissimus are you sure about the
```sh
openssl aes-256-cbc -K $encrypted_316f730cabee_key -iv $encrypted_316f730cabee_iv -in ./project/secret -out /dev/stdout -d
```
thing? are these `$encrypted_316f730cabee_key` variable available? it's complaining that [`iv undefined`](https://travis-ci.org/scalaz/scalaz-zio/jobs/418888480#L468)

**Comment by NeQuissimus** [comment]
yeah, that is what the Travis CLI gives me.

Take a look at https://gist.github.com/NeQuissimus/fe1081a66c56c26e8509f419204e5603

This is what I would expect it to look like (completely untested)

**Comment by sideeffffect** [comment]
is it the case, that `$encrypted_316f730cabee_key` and `$encrypted_316f730cabee_iv` come from the `env.global.secure` (in `.travis.yml`)? if yes, then they're available only when not building a pull-request
btw why are there 3 secrets variables?

**Comment by NeQuissimus** [comment]
I believe so.

The three secrets are `SONATYPE_USERNAME`, `SONATYPE_PASSWORD` and `PGP_PASSPHRASE`

**Comment by sideeffffect** [comment]
so in that case, `$encrypted_316f730cabee_key` and `$encrypted_316f730cabee_iv` come from Travis itself, not `.travis.yml`, right?

**Comment by NeQuissimus** [comment]
That is how I understand it

**Comment by sideeffffect** [comment]
anyway, I consider this ready to be merged, let's see how it works with publishing from `master`

**Comment by sideeffffect** [comment]
@NeQuissimus I'm sorry, it's still not working 100 %

```sh
$ export PGP_SECRET=$(openssl aes-256-cbc -K $encrypted_316f730cabee_key -iv $encrypted_316f730cabee_iv -in ./project/secret -out /dev/stdout -d)
```
is complaining with
```sh
bad decrypt
140009364244128:error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt:evp_enc.c:539:
```
see https://travis-ci.org/scalaz/scalaz-zio/jobs/418912880#L465

any idea what could be going wrong?

**Comment by sideeffffect** [comment]
there's also this error https://travis-ci.org/scalaz/scalaz-zio/jobs/418927072#L780

**Comment by sideeffffect** [comment]
and we yet have to add a tag to `master` (`v0.1.0`?) @jdegoes

**Comment by jdegoes** [comment]
I added a tag to master: `v0.2.0`. Is that enough?

**Comment by sideeffffect** [comment]
@jdegoes it should be. But we can't be sure until it runs the `ci-release` step (on `master`)
but currently `master` is broken on microsite -- can you restart it? perhaps it's an intermittent issue

**Comment by jdegoes** [comment]
@sideeffffect Restarted.

**Comment by sideeffffect** [comment]
@jdegoes :( and now it's failing in Lint
that doesn't seem to be an error on our side, Travis is not in a good shape right now

**Comment by jdegoes** [comment]
@sideeffffect JRuby is not downloading, it seems.

**Pull request merged**